src/compile.ts

```diff
@@ -407,6 +472,291 @@ export function compile(
+            undefined,
+            undefined,
+            [
+              context.factory.createParameterDeclaration(
+                undefined,
+                undefined,
+                undefined,
+                preWarmContext,
+                undefined,
+                undefined,
+                undefined
+              ),
+            ],
+            undefined,
+            undefined,
+            closure
+          ),
+          context.factory.createArrayLiteralExpression(integrations),
+        ]);
+      }
+
+      function toNativeExpr(
+        node: ts.Node,
+        context: ts.TransformationContext,
+        nativeExprContext: NativeExprContext
+      ): ts.Node | ts.Node[] | undefined {
+        if (ts.isCallExpression(node)) {
+          // Integration nodes have a static "kind" property.
+          if (isIntegrationNode(node.expression)) {
+            const outOfScopeIntegrationReference = getOutOfScopeValueNode(
+              node.expression,
+              nativeExprContext.closureNode
+            );
+
+            if (!outOfScopeIntegrationReference) {
+              throw Error(
+                "Integration Variable must be defined out of scope: " +
+                  node.expression.getText()
+              );
+            }
+
+            // add the function identifier to the integrations
+            nativeExprContext.registerIntegration(
+              outOfScopeIntegrationReference
+            );
+            // call the integration call function with the prewarm context and arguments
+            // At this point, we know native will not be undefined
+            // await integration.native.call(args, preWarmContext)
+            // FIXME: doesn't work without an array
+            // TODO: Support both sync and async function invocations: https://github.com/sam-goodwin/functionless/issues/105
+            return [
+              context.factory.createAwaitExpression(
+                context.factory.createCallExpression(
+                  context.factory.createPropertyAccessExpression(
+                    context.factory.createPropertyAccessExpression(
+                      node.expression,
+                      "native"
+                    ),
+                    "call"
+                  ),
+                  undefined,
+                  [
+                    context.factory.createArrayLiteralExpression(
+                      node.arguments
+                    ),
+                    // TODO: determine if this is needed at all?
+                    nativeExprContext.preWarmContext,
+                  ]
+                )
+              ),
+            ];
+          }
+        } else if (ts.isNewExpression(node)) {
+          const newType = checker.getTypeAtLocation(node);
+          // cannot create new resources in native runtime code.
+          const functionlessKind = getFunctionlessTypeKind(newType);
+          if (getFunctionlessTypeKind(newType)) {
```

Comment:
Should this be expanded to any class that extends Construct?

Reply:
That is what that code did below, but I was having a hard time determining if something was a Contract or not.

Reply:
Could we use structural typing heuristics? Or the fully qualified name of the type?

Reply:
Fully qualified in typescript is the path, which is different on every system, right? I wasn't able to figure out how to make good use of that. Have any examples?

Reply:
Not sure why I didn't notice this before, will give it a try.

https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/core/lib/construct-compat.ts#L21

Reply:
Got it...


```ts
      /**
       * Heuristically evaluate the fqn of a symbol to be in a module and of a type name.
       *
       * /somePath/node_modules/{module}/somePath.{typeName}
       *
       * isInstanceOf(typeSymbol, "constructs", "Construct")
       * ex: /home/sussmans/functionless/node_modules/constructs/lib/index.js
       */
      function isInstanceOf(
        symbol: ts.Symbol,
        module: string,
        typeName: string
      ) {
        const find = /.*\/node_modules\/([^\/]*)\/.*\.(.*)$/g.exec(
          checker.getFullyQualifiedName(symbol)
        );

        const [_, mod, type] = find ?? [];

        return mod === module && type === typeName;
      }

      function isCDKConstruct(type: ts.Type): boolean {
        const typeSymbol = type.getSymbol();

        return (
          ((typeSymbol &&
            isInstanceOf(typeSymbol, "constructs", "Construct")) ||
            type.getBaseTypes()?.some((t) => isCDKConstruct(t))) ??
          false
        );
      }
```